docs/cow-protocol/reference/contracts/core/vault_relayer.md

```diff
@@ -0,0 +1,99 @@
+---
+id: vault-relayer
+sidebar_position: 2
+---
+
+# GPv2VaultRelayer
+
+## Architecture
+
+The `GPv2VaultRelayer` contract is an important component used to protect user funds from malicious solvers. As previously mentioned, the `GPv2Settlement` contract allows using arbitrary on-chain liquidity through interactions (such as performing a swap on [Balancer V2](https://balancer.fi), or performing a Paraswap trade). If `Vault` and `ERC20` allowances were made directly to a `GPv2Settlement` contract, a malicious solver could drain user funds through the interaction mechanism. However, since these allowances are made to the `GPv2VaultRelayer` contract and interactions to the contract are strictly forbidden, malicious solvers have no direct access to user funds. The `GPv2Settlement` contract uses the `GPv2VaultRelayer` to withdraw user tokens only as part of the trade, which contains strong guarantees that the user's signed order parameters are respected.
+
+The `GPv2VaultRelayer` has access to user balances through 3 mechanisms:
+
+1. [Balancer External Balances](#balancer-external-balances)
+2. [Balancer Internal Balances](#balancer-internal-balances)
```

Comment:
These aren't really used in practice and should probably be explained with less priority.

Reply:
Order adjusted to take this into account.